Update vasttrafik integration to be ui configurable

zbrox opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Västtrafik integration was legacy and you couldn't configure it from the UI. With this change it is configurable from the UI and should keep backwards compatibility. 

First time contributing and I was unsure for most of the process when it comes to taking an integration out of legacy. Open to whatever feedback!

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/40566
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

zbrox committed:
update vasttrafik integration to be ui configurable

zbrox committed:
run ruff format

NoRi2909 commented on homeassistant/components/vasttrafik/strings.json:
```diff
@@ -0,0 +1,100 @@
+        "title": "Set up Västtrafik API",
+        "description": "Enter your Västtrafik API credentials. You can get these from the Västtrafik developer portal.",
+        "data": {
+          "key": "API Key",
+          "secret": "API Secret"
```
This also needs sentence-casing, and can be partly fixed by using a common string:

```suggestion
          "key": "[%key:common::config_flow::data::api_key%]",
          "secret": "API secret"
```

MartinHjelmare commented on script/hassfest/quality_scale.py:
```diff
@@ -1051,7 +1051,6 @@ class Rule:
     "uvc",
     "v2c",
     "vallox",
-    "vasttrafik",
```
Please update the quality scale record and check in a separate PR without any other changes, after this PR is merged.

zbrox committed:
add sensor tests

zbrox committed:
Update homeassistant/components/vasttrafik/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

zbrox committed:
Merge branch 'dev' into feat/vasttrafik-upgrade

zbrox committed:
restructure the config flow and entities creation

zbrox commented on homeassistant/components/vasttrafik/__init__.py:
```diff
@@ -1 +1,78 @@
+        # This is a departure board entry - get credentials from main integration
+        main_entry = _get_main_entry(hass)
+        if not main_entry:
+            raise ConfigEntryNotReady("Main Västtrafik integration not found")
```
I wasn't entirely sure about some details but this is how it currently looks 
<img width="2118" height="1180" alt="CleanShot 2025-08-21 at 16 51 37@2x" src="https://github.com/user-attachments/assets/d8eac282-ce7f-4968-a927-c7854acbff08" />

gjohansson-ST reviewed (changes requested):
Some initial comments where-of some might be a bit big so feel free to ask if it's not entirely clear.
Besides that you have subentries, there should be plenty of previous PR's to look at.

zbrox committed:
Apply suggestions from code review

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Add single_config_entry and fix config flow strings

- Add single_config_entry: true to manifest to enforce single API config entry
- Add data/data_description fields to user step in strings.json
- Use common string reference for API key
- Fix sentence casing: 'API secret' (lowercase)

zbrox committed:
Add complete type hints to config flow functions

- Add ConfigEntry type hint to search_stations config_entry parameter
- Add complete return type: tuple[list[dict[str, str]], str | None]
- Add complete signature to async_get_supported_subentry_types
- Remove unused typing.Any import from sensor.py

zbrox committed:
Remove self-explanatory comments

zbrox committed:
Improve error handling and fix mandatory field access

zbrox committed:
Replace @Throttle with SCAN_INTERVAL constant

zbrox committed:
Add YAML configuration import support

zbrox committed:
Add update listener to reload entry when subentries change

zbrox committed:
Use translations for entity names instead of hardcoded strings

zbrox committed:
Add comprehensive YAML import tests

Tests cover:
- Import with no departures
- Import with multiple departures creating subentries
- Single instance enforcement during import
- Exception handling (InvalidAuth, CannotConnect, unknown errors)
- YAML sensor platform import with repair issue creation

Follows Nederlandse Spoorwegen PR pattern

zbrox commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -127,17 +237,28 @@ def native_value(self):
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self) -> None:
+    async def async_update(self) -> None:
```
After reading this part of the docs https://developers.home-assistant.io/docs/integration_fetching_data/ it seems this is more fitted for the SCAN_INTERVAL approach. Since we are sending quite different requests - depending on the configured departure board - we cannot use the HA coordinator utils.

zbrox commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,429 @@
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        # Check if we already have a Västtrafik integration configured
+        existing_entries = self._async_current_entries(include_ignore=False)
```
Thanks, I have missed this very obvious thing :) It's added to the manifest

zbrox commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -66,42 +69,154 @@ def setup_platform(
-        self._name = name or departure
-        self._departure = self.get_station_id(departure)
-        self._heading = self.get_station_id(heading) if heading else None
+        self._attr_name = name or f"{departure} Departures"
```
This should be using strings with placeholders now

zbrox commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -66,42 +69,154 @@ def setup_platform(
-        (
+
+    # Create departure sensors - no device, entities appear under service directly
+    sensors = []
```
I wasn't sure what to do with them but I have found #151567 and saw how it was done there

zbrox committed:
Merge branch 'dev' into feat/vasttrafik-upgrade

emontnemery commented on requirements_test_all.txt:
```diff
@@ -2557,6 +2557,15 @@ volvocarsapi==0.4.3
+
+# homeassistant.components.vultr
+vultr==0.1.2
+
```
Why are these added? Is it a merge mistake?

zbrox replied:
Oh yeah, this seems like a merge mistake that I have overlooked. Sorry, I will remove those.

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+
+        if stations:
+            return stations, None
+        return [], "no_stations_found"
```
We want only that which can actually raise to be in the try-clause.
Move code which can't raise to an else clause, or to after the try-except

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+                STEP_USER_DATA_SCHEMA, suggested_values
+            ),
+            errors=errors,
+        )
```
I'd suggest to simplify:
```suggestion
        return self.async_show_form(
            step_id="reconfigure",
            data_schema=self.add_suggested_values_to_schema(
                STEP_USER_DATA_SCHEMA, user_input or entry.data
            ),
            errors=errors,
        )
```

zbrox replied:
Makes sense

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+            else:
+                parent_entry = self._get_entry()
+                if not parent_entry:
+                    return self.async_abort(reason="parent_entry_not_found")
```
When does this happen?

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+        if user_input is not None:
+            selected_station = user_input.get("station")
+            if not selected_station:
+                errors["base"] = "station_required"
```
Is this possible?

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+                    track.strip()
+                    for track in user_input[CONF_TRACKS].split(",")
+                    if track.strip()
+                ]
```
Instead of asking the user to input comma-separated lists of tracks and lines, can we use the textselector in multiple mode?

zbrox replied:
Yeah, I was thinking of doing that but eventually it was so much other things I never got around to it. I will look into it.

emontnemery commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+            description_placeholders={
+                "station_name": self._selected_station,
+                "lines_help": "Enter line numbers separated by commas (e.g. 1, 2, 55)",
+                "tracks_help": "Enter track/platform numbers separated by commas (e.g. A, B, 1, 2)",
```
What's this?

emontnemery reviewed (changes requested):
I think this looks good overall!
Some comments though. I've not reviewed the changes in sensor or the tests yet.

zbrox commented on homeassistant/components/vasttrafik/manifest.json:
```diff
@@ -1,10 +1,13 @@
   "quality_scale": "legacy",
-  "requirements": ["vtjp==0.2.1"]
+  "requirements": ["vtjp==0.2.1"],
+  "single_config_entry": true
```
We use only one config entry where we keep the API credentials and that is it. There is no real world use case where we would want to have 2 API credential sets really. Not that I can think of at least. So you configure once and create entities with that one config. You can reconfigure later if keys expire and such.

zbrox commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -12,35 +12,38 @@
+_LOGGER = logging.getLogger(__name__)
 
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=120)
+SCAN_INTERVAL = timedelta(seconds=60)
```
Yeah, I was wondering and decided that one minute is probably a good interval

emontnemery replied:
OK, that's probably fine, but let's not do that in a PR which is about adding a config flow.

zbrox commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+        errors = {}
+
+        if user_input is not None:
+            selected_station = user_input.get("station")
```
This is more of a reflex on my end, using get always as a defensive approach. I can change that.

zbrox committed:
Remove packages due to a merge mistake when syncing fork

zbrox committed:
Remove unnecessary checks in select station step

zbrox committed:
Remove artificial limit on station search results

zbrox committed:
Remove another unnecessary check leftover from previous iterations

zbrox committed:
Fix try/catch to only surround code which can throw errors

zbrox committed:
Keep user input on reconfigure step

zbrox committed:
Some formatting

zbrox committed:
Remove unnecessary check in initial setup

zbrox committed:
Refactor validate_api_credentials to return errors instead of raising

Simplifies the usage of the validate_api_credentials function

zbrox committed:
Rename to more fit to its purpose name

zbrox committed:
Remove now unused and empty options flow

zbrox committed:
Remove artificial filter on number of departues stored

Maybe add an optional configuration for it in the departure board
configuration

zbrox committed:
Remove leftover hardcoded placeholders

zbrox committed:
Simplify the departure (re)configure steps

Reuse the same schema
Use text selector for simpler input of multivalue filters

emontnemery commented:
@zbrox please avoid force-pushing. Instead merge upstream/dev to your branch, either manually or by using this button: 
<img width="168" height="47" alt="image" src="https://github.com/user-attachments/assets/79338384-5de1-4b41-8fcc-be76738342ee" />

zbrox commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,459 @@
+                vol.Optional(CONF_HEADING): str,
+                vol.Optional(CONF_LINES): str,
+                vol.Optional(CONF_TRACKS): str,
+                vol.Optional(CONF_DELAY, default=DEFAULT_DELAY): vol.Coerce(int),
```
I used the text selector with the multiple flag on for lines and tracks. Should the rest be also text selectors like name, heading since they are just a very simple case of any text goes? Also should the delay remain as is?

zbrox committed:
More mypy fixes, somehow missed before

zbrox committed:
Fix tests after previous refactor

zbrox committed:
Use SelectSelector for station selection

zbrox committed:
Fix errors reported by mypy, ruff, pylint

zbrox committed:
Run prettier --write to fix json not conforming to the style guidelines

zbrox committed:
Fix of the tests forgotten to be commited previously

zbrox committed:
Run hassfest

gjohansson-ST commented on homeassistant/components/vasttrafik/__init__.py:
```diff
@@ -1 +1,51 @@
+
+async def async_reload_entry(hass: HomeAssistant, entry: VasttrafikConfigEntry) -> None:
+    """Reload entry when subentries change."""
+    await hass.config_entries.async_reload(entry.entry_id)
```
```suggestion
```

gjohansson-ST replied:
```suggestion
    hass.config_entries.async_schedule_reload(entry.entry_id)
```

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,350 @@
+                return self.async_create_entry(title="Västtrafik", data=user_input)
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors or {}
```
```suggestion
            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
```

zbrox replied:
Mypy is complaining if we remove the default value here. Since above the errors can be reset to be a None

gjohansson-ST replied:
Sorry, didn't check but default `error` to be an empty dict and not `None`

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,350 @@
+
+        station_schema = vol.Schema(
+            {
+                vol.Required("station"): vol.In(station_options),
```
If you use the `SelectSelector`instead this could be presented nicely with the proper labels instead of just the station value.

zbrox committed:
Remove unnecessary code due to single config entry

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Use async_schedule_reload instead of async_reload

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Run ruff format

zbrox committed:
Revert single_config_entry

zbrox committed:
Create issues when importing the YAML config

zbrox committed:
Remove unnecessary check for the subentry type

There is only the departure board type

zbrox committed:
Explicitly convert dicts to StationData

zbrox committed:
Restructure and simplify the structure of the integration after review
comments

- Replace station name lookups with station GIDs stored during the
  config flow
	- we store the GID and we don't need to look it up on startup
	  anymore
	- we do the lookup only during the yaml import now
	- we remove heading matching through the API and move it in the
	  filtering like the rest of the filters (lines, tracks)
- Simplify the unique id and just use subentry ID as the unique ID
	- this also enables several departure boards per station, which
	  was not possible before
- Original board name is now only used as a placeholder/suggested value, 
which the user can overwrite and only the custom name is passed to 
the sensor since we now use the GID
- add proper type hints to the init of VasttrafikDepartureSensor
- improve debug logging in data fetching by including the heading filter

zbrox committed:
Remove unnecessary creation of issues on import

zbrox committed:
During reconfigure check that credentials are not repeated

zbrox committed:
Fix some more exception management

zbrox committed:
Use a constant for the search query field name instead of a hardcoded
string

zbrox committed:
Return empty dict for no errors instead of None

zbrox committed:
Use Selectors in the schemas for all fields

zbrox committed:
Remove default value which was wrong, as per review suggestion

zbrox committed:
Improve and simplify the issues created

zbrox committed:
Fix some ruff errors when catching an exception

zbrox committed:
Remove check for search being too short

Schema requirement will guarantee that

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,351 @@
+        errors: dict[str, str] | None = {}
+        if user_input is not None:
+            if not (errors := await validate_api_credentials(self.hass, user_input)):
+                return self.async_update_reload_and_abort(
```
```suggestion
                return self.async_update_and_abort(
```
This should not reload since you have an update listener which does it

gjohansson-ST commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -60,49 +67,150 @@
+        self._attributes: dict[str, Any] | None = None
+
+        # Create device info for departure board subentries (similar to Ollama pattern)
+        if subentry_id:
```
This is unneccessary as you have a subentry_id (which I suggest to remove as you don't need the check at all).

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,351 @@
+
+async def search_stations(
+    hass: HomeAssistant, config_entry: ConfigEntry, query: str
+) -> tuple[list[SelectOptionDict], str | None]:
```
You should replace `SelectOptionDict` with `dict[str, str]` as you don't use `SelectOptionDict` anywhere to create the dictionary. That typing isn't correct.

zbrox commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,350 @@
+            data_schema=self.add_suggested_values_to_schema(
+                STEP_USER_DATA_SCHEMA, user_input or entry.data
+            ),
+            errors=errors or {},
```
Ah, silly me, confused it a bit

zbrox commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,351 @@
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+
+        errors: dict[str, str] | None = {}
```
Yeah, I just realised it as I was checking if I'd miss a comment 🤦 Confused this with the other one where it is actually passed to show form for example where it definitely should be None and not `or {}` as I had it

zbrox committed:
Update homeassistant/components/vasttrafik/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Update homeassistant/components/vasttrafik/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Update homeassistant/components/vasttrafik/sensor.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

zbrox committed:
Use station GID as a selector value

zbrox commented on homeassistant/components/vasttrafik/sensor.py:
```diff
@@ -60,49 +67,150 @@
+
+            device_name = " • ".join(device_name_parts)
+
+            self._attr_device_info = DeviceInfo(
```
I started doing that but we use `self._attr_name` in a lot of logger calls throughout the class and it makes it a pain if we change to use `self._attr_device_info.name` since by type it's optional. So maybe better leave it as is?

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,384 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_KEY): str,
```
```suggestion
        vol.Required(CONF_KEY): TextSelector(),
```
Use the selectors on all fields (it was mentioned in some previous comment from emontnemery)

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,384 @@
+            return {"base": "invalid_auth"}
+        return {"base": "cannot_connect"}
+    except Exception as err:  # noqa: BLE001
+        _LOGGER.error("Unexpected error validating Västtrafik credentials: %s", err)
```
```suggestion
        _LOGGER.exception("Unexpected error validating Västtrafik credentials: %s", err)
```

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,384 @@
+        _LOGGER.error("Unexpected error validating Västtrafik credentials: %s", err)
+        return {"base": "unknown"}
+
+    return None
```
```suggestion
    return {}
```

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,384 @@
+
+        errors: dict[str, str] | None = {}
+        if user_input is not None:
+            if not (errors := await validate_api_credentials(self.hass, user_input)):
```
Missing a check here that the reconfigure isn't using an api key already in use (given that we allow for multiple config entries as earlier discussed)

gjohansson-ST commented on homeassistant/components/vasttrafik/config_flow.py:
```diff
@@ -0,0 +1,384 @@
+    ) -> SubentryFlowResult:
+        """Handle the initial step of adding a departure board."""
+        if user_input is not None:
+            search_query = user_input.get("search_query", "").strip()
```
```suggestion
            search_query = user_input["search_query"].strip()
```
Create a constant for `search_query` and use that

gjohansson-ST reviewed (changes requested):
A lot of nitpicks in the config flow. Most of it mentioned before but you have resolved those comments without actually fixing them.

zbrox committed:
Update homeassistant/components/vasttrafik/config_flow.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>
